presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java

```diff
@@ -46,10 +49,13 @@
     private final Optional<Integer> hashChannel;
     private final OperatorContext operatorContext;
     private final long memorySizeBeforeSpill;
+    private final long memoryLimitForMergeWithMemory;
     private Optional<Spiller> spiller = Optional.empty();
     private Optional<MergingHashAggregationBuilder> merger = Optional.empty();
     private CompletableFuture<?> spillInProgress = CompletableFuture.completedFuture(null);
-    private final LocalMemoryContext systemMemoryContext;
+    private final LocalMemoryContext aggregationMemoryContext;
```

Comment:
you could make two separate commits (both for better clarity and for better stats on github :) ). One refactoring those contexts and second adding merge with memory